### PR TITLE
Official ARM builds for every new release

### PR DESCRIPTION
**Pull request by luxas**
This PR features support for building `arm`(, `arm64` and `ppc64le`) binaries dynamically or statically on `amd64` hosts. It builds docker images for `kube-apiserver`, `kube-controller-manager`, `kube-scheduler`, `kube-proxy` for the new architectures and pushes them. 
Bug fixed: `hyperkube` (legacy name for `hyperkube-amd64`) is pushed too when releasing for compability with older setups

Earlier: 
- `arm`
  - Server support only when building **on** an ARM hosts
  - Client support also on `amd64` hosts
- No `arm64` or `ppc64le` support at all.

Now:
- `arm`(, `ppc64le`, `arm64` when we have `go1.5`)
  - Server support when building on `amd64` hosts
    - `kubelet` and `hyperkube` especially are built dynamically to include `cAdvisor`
  - One patch already merged: #19712 for `arm64` to build server targets
  - Client support present
  - Test target support fixed for `arm64`: onsi/ginkgo#209 and #20546 
  - Test target support isn't so important for `arm`, `arm64`, `ppc64le`, so test targets are only built for `linux/amd64`, `darwin/amd64` and `windows/amd64`.
- `windows/386` client support added

Tracking issue: #17981 
Depends on: #14873 
A bit related to: #19703 

When this is merged, it has the advantage that builds will fail when trying to merge only-working-on-amd64 code for the server targets, so it's easily noticed.

This is WIP for now because #14873 hasn't merged yet (and I have to rebase after that)

> ~~However, if it takes too long time for  `arm64` and `ginkgo` to work, I'm gonna comment out `arm64` for the moment and merge that later.~~

Thoughts? This will make Kubernetes more cross-platform, and that will be awesome :+1: 

@brendandburns @thockin @zmerlynn @fgrzadkowski @david-mcmahon @mikedanese @gmarek @wojtek-t @qq511700505

### CODE REVIEW

**Comment by zmerlynn** [comment]
@k8s-bot ok to test

File: build/build-image/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2014 Google Inc. All rights reserved.
+# Copyright 2016 Google Inc. All rights reserved.
```
Comment (zmerlynn) [inline build/build-image/Dockerfile]: Please don't bump the copyright on the file unless it's been basically rewritten (it hasn't).

File: build/build-image/cross/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2014 Google Inc. All rights reserved.
+# Copyright 2016 Google Inc. All rights reserved.
```
Comment (zmerlynn) [inline build/build-image/cross/Dockerfile]: Ditto.

File: hack/lib/golang.sh [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-# Copyright 2014 The Kubernetes Authors All rights reserved.
+# Copyright 2016 The Kubernetes Authors All rights reserved.
```
Comment (zmerlynn) [inline hack/lib/golang.sh]: Ditto

File: build/build-image/Dockerfile (hunk 1 above)
Comment (luxas) [inline build/build-image/Dockerfile]: Aha, thought it could be updated... No problem

**Comment by luxas** [comment]
@zmerlynn Do you have any specific reason why you are building test targets for all client platforms? Isn't that a little bit unnecessary and slower? 
I'm wondering if it could be removed from `hack/build-cross.sh`... (That code was modified last time 2014)

Of course test targets should be made for `amd64`, but for `arm`, `windows` and `ppc64le` for example it seems a little bit strange

I know that it might be another discussion though and that it might break backwards-compability (I don't know if you run integration tests on windows)
So I'm just asking...

**Comment by ixdy** [comment]
FYI also @ihmccreery @david-mcmahon

**Comment by zmerlynn** [comment]
@luxas: Yes, we could probably take the test code out of the client build.

File: hack/lib/golang.sh [hunk 4]
```diff
@@ -97,15 +99,20 @@ readonly KUBE_TEST_PORTABLE=(
...
 readonly KUBE_CLIENT_PLATFORMS=(
   linux/amd64
   linux/386
   linux/arm
+#  linux/arm64
+  linux/ppc64le
   darwin/amd64
   darwin/386
   windows/amd64
+  windows/386
```
Comment (zmerlynn) [inline hack/lib/golang.sh]: Did you add 32-bit windows for a reason?

**Comment by zmerlynn** [comment]
This looks okay. I'm a little concerned about bumping `golang` given what lead up to #19487, so I'm not sure how to adjudicate that.

**Comment by ixdy** [comment]
#17524 I think you mean?

In any case, it'd probably be good to bump the golang version separate from other changes. #14873 is for that, I think.

**Comment by zmerlynn** [comment]
Yes, I have no idea where I pulled the other number from.

File: hack/lib/golang.sh (hunk 4 above)
Comment (luxas) [inline hack/lib/golang.sh]: Because I saw that a compiler in the earlier `kube-build:cross` was built for `windows/386`, but it wasn't listed as a client platform. Supposed it just had been forgotten.

But it doesn't harm to build `kubectl` for 32-bit win also.

**Comment by luxas** [comment]
`golang` should be updated in another PR, probably #14873.
I just have to include it now in this WIP stage so it builds.
Will rebase when #14873 gets in

Will rebase and rework the PR...

File: build/common.sh [hunk 5]
```diff
@@ -97,6 +97,7 @@ readonly GCS_STAGE="${LOCAL_OUTPUT_ROOT}/gcs-stage"
 # The set of master binaries that run in Docker (on Linux)
 # Entry format is "<name-of-binary>,<base-image>".
 # Binaries are placed in /usr/local/bin inside the image.
+# TODO: Make base images for multiple server platforms
```
Comment (luxas) [inline build/common.sh]: @zmerlynn Do you have some suggestions how to nicely handle this?
We need to be able to specify different base images for different arches

**Comment by luxas** [comment]
Looks like a e2e flake. 
It built binaries and packaged them as it should. 

But some test didn't start in time or something...

File: build/build-image/cross/Dockerfile [hunk 6]
```diff
@@ -15,16 +15,22 @@
 # This file creates a standard build environment for building cross
 # platform go binary for the architecture kubernetes cares about.
 
-FROM  golang:1.4.2
+FROM  golang:1.5.3
```
Comment (thockin) [inline build/build-image/cross/Dockerfile]: @ihmccreery @lavalamp is this going to be a problem?

File: build/common.sh [hunk 7]
```diff
@@ -796,7 +798,7 @@ function kube::release::create_docker_images_for_server() {
         ln ${binary_dir}/${binary_name} ${docker_build_path}/${binary_name}
         printf " FROM ${base_image} \n ADD ${binary_name} /usr/local/bin/${binary_name}\n" > ${docker_file_path}
 
-        local docker_image_tag=gcr.io/google_containers/$binary_name:$md5_sum
+        local docker_image_tag=gcr.io/google_containers/${binary_name}-${arch}:${md5_sum}
```
Comment (thockin) [inline build/common.sh]: will we undo this naming convention if/when Docker supports arch images?

File: build/build-image/cross/Dockerfile (hunk 6 above)
Comment (luxas) [inline build/build-image/cross/Dockerfile]: No, I've bumped it just so this PR builds. 
As I said above, I will rebase this one after something like #14873 is merged.

File: build/common.sh (hunk 7 above)
Comment (luxas) [inline build/common.sh]: Probably, but docker is far away from that right now.
@fgrzadkowski already fixed it and the naming convention is `binary-arch:version`
It's just that he forgot to fix the development image. Some lines down, he has fixed it already for the release images

**Comment by luxas** [comment]
@zmerlynn Rebased to master, now when we have `go1.5.3` support OOTB.
The next thing to do is to use right base images when cross-building. Right now, the cross-compiled binary is put into a docker image e.g. `FROM busybox`. We need to have arch-specific images there.
Ideas how to implement?

BTW, do you know why the bot downgraded this to a size/XS PR and added kind/design?
When I remove `WIP`, I'll squash all commits into one of course

**Comment by ixdy** [comment]
ok to test

**Comment by luxas** [comment]
@ixdy Are there some problems with the bot or is it normal that is periodically requires "ok to test"?

And do you have any clue why the build fails? The strange thing is, that building succeeds, but kube-up fails, always on the same step. I don't know if it's related to this or some other PR or if it's an internal bug.

**Comment by ixdy** [comment]
@luxas I updated the GitHub pull request Jenkins plugin yesterday, and the new version apparently forgot basically everything it knew about PRs, unfortunately. Hopefully this won't happen again.

**Comment by ixdy** [comment]
@k8s-bot e2e test this again please

**Comment by luxas** [comment]
I would like to squash this into one commit and fix the base image problem before this gets merged though (hopefully today)
Are you OK with `linux/amd64`, `darwin/amd64` and `windows/amd64` as test targets? More or less platforms?

The `onsi/ginkgo` `linux/arm64` problem is fixed: onsi/ginkgo#209 on master. 
Maybe we could update the Godep...

**Comment by ixdy** [comment]
Should probably update the Ginkgo godep in a separate commit at least, maybe a separate PR entirely.

**Comment by ixdy** [comment]
e2e keeps failing to initialize the cluster, so something is broken here.

**Comment by luxas** [comment]
@ixdy Yes, `ginkgo` is a separate PR.
Do you have any idea why it fails?

**Comment by ixdy** [comment]
Unfortunately not. #10423 (assigned to me) still hasn't been fixed.

@k8s-bot e2e test this

I'll try to log in while it's running to see if there are any useful logs there.

**Comment by ixdy** [comment]
The `docker.log` on the master from the most recent run (https://gist.github.com/ixdy/f053757dd90bdb029874) looks suspicious, though I don't know to better understand what's happening:
```
time="2016-01-29T22:15:24.147859469Z" level=error msg="HTTP Error" err="No such image: gcr.io/google_containers/kube-apiserver:263e0cee7786f2228dd7c9292da4a57d" statusCode=404 
time="2016-01-29T22:15:34.144886463Z" level=error msg="Handler for GET /images/{name:.*}/json returned error: No such image: gcr.io/google_containers/kube-controller-manager:6f43727c1a9650b85edc7525c0edca6e" 
time="2016-01-29T22:15:34.144940873Z" level=error msg="HTTP Error" err="No such image: gcr.io/google_containers/kube-controller-manager:6f43727c1a9650b85edc7525c0edca6e" statusCode=404 
time="2016-01-29T22:15:35.146185334Z" level=error msg="Handler for GET /images/{name:.*}/json returned error: No such image: gcr.io/google_containers/kube-apiserver:263e0cee7786f2228dd7c9292da4a57d" 
time="2016-01-29T22:15:35.146229175Z" level=error msg="HTTP Error" err="No such image: gcr.io/google_containers/kube-apiserver:263e0cee7786f2228dd7c9292da4a57d" statusCode=404 
time="2016-01-29T22:15:35.146982071Z" level=error msg="Handler for GET /images/{name:.*}/json returned error: No such image: gcr.io/google_containers/kube-scheduler:32e7cbf068ca07f3215aa8c9e50dda70" 
time="2016-01-29T22:15:35.147012434Z" level=error msg="HTTP Error" err="No such image: gcr.io/google_containers/kube-scheduler:32e7cbf068ca07f3215aa8c9e50dda70" statusCode=404 

```

**Comment by ixdy** [comment]
Added the master's kubelet.log to that gist, too.

**Comment by luxas** [comment]
Ah, of course. Thanks @ixdy. 
I think I may fix the issue. Sending a new PR soon

File: build/common.sh [hunk 8]
```diff
@@ -1480,6 +1525,16 @@ function kube::release::docker::release() {
       local docker_target="${KUBE_DOCKER_REGISTRY}/${binary}-${arch}:${KUBE_DOCKER_IMAGE_TAG}"
       kube::log::status "Pushing ${binary} to ${docker_target}"
       "${docker_push_cmd[@]}" push "${docker_target}"
+
+      # If we have a amd64 docker image. Tag it without -amd64 also and push it for compability with earlier versions
+      if [[ ${arch} == "amd64" ]]; then
```
Comment (luxas) [inline build/common.sh]: @fgrzadkowski PTAL
Fixes: https://github.com/kubernetes/kubernetes/pull/19061#discussion_r50636543

File: build/common.sh [hunk 9]
```diff
@@ -793,7 +828,14 @@ function kube::release::create_docker_images_for_server() {
         ln ${binary_dir}/${binary_name} ${docker_build_path}/${binary_name}
         printf " FROM ${base_image} \n ADD ${binary_name} /usr/local/bin/${binary_name}\n" > ${docker_file_path}
 
-        local docker_image_tag=gcr.io/google_containers/$binary_name:$md5_sum
+        if [[ ${arch} == "amd64" ]]; then
```
Comment (luxas) [inline build/common.sh]: @ixdy The issue was probably here. Some setup scripts assumed that the name of `linux/amd64` images was `${binary_name}:${md5_sum}` while I had changed it to  `${binary_name}-${arch}:${md5_sum}` when I was thinking about other arches.

This is the way to go. When we're releasing, both kinds of `amd64` images will be pushed

**Comment by luxas** [comment]
Some numbers with this patch: 
`kube-build:cross`: ~12 min
server platforms: `linux/amd64` `linux/arm` **`linux/arm64` `linux/ppc64le`**: ~4 min
client platforms: `linux/amd64` `linux/386` `linux/arm` **`linux/arm64` `linux/ppc64le`** `darwin/amd64` `darwin/386` `windows/amd64` **`windows/386`**: ~1 min
test platforms: `linux/amd64` `darwin/amd64` `windows/amd64`: ~2 min
packaging: ~4 min

So I think build performance is OK

File: build/common.sh (hunk 9 above)
Comment (ixdy) [inline build/common.sh]: which setup scripts are those? seems like maybe we should update them to always use `${arch}` in the name?

Comment (luxas) [inline build/common.sh]: At least something here: https://github.com/kubernetes/kubernetes/blob/master/cluster/common.sh#L400

and maybe related to https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/install.sh#L27

I think there are more places too, but can we take that in a separate PR or something?
To adopt `-amd64` naming everywhere is separate from building for other arches

**Comment by luxas** [comment]
@zmerlynn @mikedanese @ixdy Do you think we can merge this now?

File: build/common.sh (hunk 9 above)
Comment (ixdy) [inline build/common.sh]: Separate PR sounds good.

**Comment by zmerlynn** [comment]
@luxas: Sorry for the delay, I've been swamped. I'll try to get to this soon.

File: build/common.sh [hunk 10]
```diff
@@ -90,17 +90,48 @@ readonly RELEASE_STAGE="${LOCAL_OUTPUT_ROOT}/release-stage"
...
-readonly KUBE_DOCKER_WRAPPED_BINARIES=(
-  kube-apiserver,busybox
-  kube-controller-manager,busybox
-  kube-scheduler,busybox
-  kube-proxy,gcr.io/google_containers/debian-iptables:v1
-)
+# 
+# $1 - server architecture
+kube::build::get_docker_wrapped_binaries() {
+  if [[ $1 == "amd64" ]]; then
```
Comment (zmerlynn) [inline build/common.sh]: Any reason not to use a `case` statement here?

File: build/build-image/cross/Dockerfile [hunk 11]
```diff
@@ -40,3 +48,11 @@ RUN apt-get install -y g++ && apt-get clean && rm -rf /var/lib/apt/lists/* &&\
...
     protoc --version
 
+# Use dynamic cgo linking for architectures other than amd64 for the server platforms
+# More info here: https://wiki.debian.org/CrossToolchains
+RUN echo "deb http://emdebian.org/tools/debian/ jessie main" > /etc/apt/sources.list.d/cgocrosscompiling.list \
+  && curl -s http://emdebian.org/tools/debian/emdebian-toolchain-archive.key | apt-key add - \
+  && for platform in ${KUBE_DYNAMIC_CROSSPLATFORMS}; do dpkg --add-architecture ${platform}; done \
+  && apt-get update \
+  && apt-get install -y build-essential \
+  && for platform in ${KUBE_DYNAMIC_CROSSPLATFORMS}; do apt-get install -y crossbuild-essential-${platform}; done
```
Comment (zmerlynn) [inline build/build-image/cross/Dockerfile]: nit: missing newline

File: build/common.sh [hunk 12]
```diff
@@ -49,7 +49,7 @@ readonly KUBE_GCS_DELETE_EXISTING="${KUBE_GCS_DELETE_EXISTING:-n}"
 # Constants
 readonly KUBE_BUILD_IMAGE_REPO=kube-build
 readonly KUBE_BUILD_GOLANG_VERSION=1.5.3
-readonly KUBE_BUILD_IMAGE_CROSS_TAG="cross-${KUBE_BUILD_GOLANG_VERSION}-1"
+readonly KUBE_BUILD_IMAGE_CROSS_TAG="cross-${KUBE_BUILD_GOLANG_VERSION}-2"
```
Comment (zmerlynn) [inline build/common.sh]: thanks for remembering :)

File: build/common.sh (hunk 10 above)
Comment (luxas) [inline build/common.sh]: One simple reason, I haven't used it before :)
At least not that much.
I'll check the syntax...

File: build/build-image/cross/Dockerfile (hunk 11 above)
Comment (luxas) [inline build/build-image/cross/Dockerfile]: Will fix, but may I ask why? Just for curiosity...

File: build/common.sh (hunk 12 above)
Comment (luxas) [inline build/common.sh]: Yeah, noticed it when I read through the file :)

File: build/common.sh (hunk 10 above)
Comment (luxas) [inline build/common.sh]: Fixed

**Comment by luxas** [comment]
@zmerlynn Updated the PR. OK to merge?

**Comment by luxas** [comment]
Unfortunately something has changed in the source now, so `arm64` is failing.
`bolt` Godep needs to be updated. I'll send a separate PR if I get godep working.

The funny thing is, the Godep is the same as when e2e succeeded earlier, so it must have been unused or something earlier...

A good thing is, when this PR is merged, e2e will catch all these bugs

File: build/build-image/cross/Dockerfile (hunk 11 above)
Comment (zmerlynn) [inline build/build-image/cross/Dockerfile]: GitHub nags :)

Comment (zmerlynn) [inline build/build-image/cross/Dockerfile]: (As do editors - emacs would actually just fix it if I loaded this, resulting in unnecessary diff the next time.)

**Comment by fgrzadkowski** [comment]
@k8s-bot add to whitelist

**Comment by luxas** [comment]
Thanks @fgrzadkowski!

@zmerlynn @mikedanese @ixdy 
I'll rebase this one after #20546 is merged and use `go1.4.2` instead of `go1.5.3` because of the regressions described in #20577

**Comment by luxas** [comment]
@ixdy @fgrzadkowski @mikedanese @zmerlynn 
Rebased to #20546 and #20577 
Is it OK to merge when green?

File: build/common.sh (hunk 8 above)
Comment (luxas) [inline build/common.sh]: ping @fgrzadkowski

**Comment by luxas** [comment]
To rebase to `go1.4` was an attempt to do this anyway, and see if it works, but it didn't.
`linux/arm64` and `linux/ppc64le` was introduced in `go1.5`

I'll try to get working this only for `linux/arm`, since that's the primary goal for this patch
We'll continue with the others `post-v1.2` when we have `go1.5.3` (or `go1.6`?) from the beginning

**Comment by zmerlynn** [comment]
LGTM

**Comment by luxas** [comment]
Thanks @zmerlynn!
BTW, could we add a release note for this? e.g.
> Official ARM builds for every release (#19769, @luxas)

I think it would be great to announce that we now support ARM :)
(If I remember correctly there's a label for release notes)

**Comment by david-mcmahon** [comment]
@luxas To add this as a release note, add the release-note label and update the PR description to whatever you want the release note to say.

**Comment by mikedanese** [comment]
@david-mcmahon adding a label requires repo write access which not everyone has

**Comment by david-mcmahon** [comment]
@mikedanese thanks for the clarification.  Adding release-note label.  @luxas please adjust release note to taste.
